### PR TITLE
CP-44531 Toolstack: Add dom0 CPU usage, total and free memory into rrdd

### PR DESCRIPTION
**Pull request by LunfanZhang**
This PR aims to add the Dom0 CPU usage and free memory into rrdd:

- For the Dom0 CPU usage, I extracted and converted it from _xenctrl.domain_getinfo.cpu_time_, I have maintained special treatment for **uptime** like we handled it in the vCPU usage since I am not sure whether we will encounter the same issue as the _vcpuinfo cputime_. This change will update the CPU usage for all domains, not just Dom0.
- For the free memory, I did not find the domain free memory related attribute in _xenctrl_, so I extracted it from '/proc/meminfo'  using **memory available**, which is the same as the guest free memory.

New test result:
```
xe vm-data-source-query uuid=0907edb2-50f8-4784-9971-38c3102337c5 data-source=cpu_usage
0.046414
```
```
xe vm-data-source-query uuid=0907edb2-50f8-4784-9971-38c3102337c5 data-source=memory_internal_free
2256969216.000000
```

### CODE REVIEW

File: ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml [hunk 1]
```diff
@@ -237,57 +237,68 @@ let dss_vcpus xc doms =
...
-                 ~description:
-                   "Fraction of time that some VCPUs are runnable and some are \
-                    blocked"
-                 ~ty:Rrd.Derive ~default:false ~min:0.0 ()
-             )
-          :: dss
-        with _ -> dss
+        let dom_cpu_time =
+            Int64.(
+              to_float @@ logand dom.Xenctrl.cpu_time xen_flag_complement
```
Comment (psafont) [inline ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml:242]: Why is this needed? This is collected per cpu in the precious lines

File: ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml [hunk 2]
```diff
@@ -237,57 +237,68 @@ let dss_vcpus xc doms =
...
+                  ~units:"(fraction)"
+                  ~value:(Rrd.VT_Float (Int64.to_float ri.Xenctrl.time5 /. 1.0e9))
+                  ~description:
+                    "Fraction of time that some VCPUs are runnable and some are \
+                      blocked"
+                  ~ty:Rrd.Derive ~default:false ~min:0.0 ()
+              )
+            :: ( Rrd.VM uuid
+              , Ds.ds_make ~name:(Printf.sprintf "cpu_usage") ~units:"(fraction)"
+                  ~description:(Printf.sprintf "domain: %d CPU usage" domid)
```
Comment (psafont) [inline ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml:297]: The rest of the metrics avoid mentioning domains, probably because it leaks abstraction. For example, whenever a VM migrates it will the description will change, this can affect users consolidating stats

File: ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml (hunk 1 above)
Comment (LunfanZhang) [reply ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml:242]: I noticed the per vcpu retains this special process, I am not sure whether the per domain will encounter this issue or not so I kept it, if we can ensure per domain is ok then I will delete it.

File: ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml (hunk 2 above)
Comment (LunfanZhang) [reply ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml:297]: That make sense!  I will delete the domain id from description.

**Review by psafont — APPROVED** [review]

File: ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml [hunk 3]
```diff
@@ -545,6 +557,29 @@ let bytes_per_mem_vm = 1024
 
 let mem_vm_writer_pages = ((max_supported_vms * bytes_per_mem_vm) + 4095) / 4096
 
+let read_mem_available () =
+  let meminfo_file = "/proc/meminfo" in
```
Comment (lindig) [inline ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml:561]: This could be a simpler. Below is code using `Scanf` and using the error monad (which you probably don't want and stick with exceptions). I have converted kB to bytes but maybe sticking with kB is fine, too.

```
let error fmt = Printf.kprintf (fun msg -> Result.error (`Msg msg)) fmt

let ok x = Result.ok x

let ( let* ) = Result.bind

let finally f always = Fun.protect ~finally:always f

let scanning path f =
  let io = Scanf.Scanning.open_in path in
  finally (fun () -> f io) (fun () -> Scanf.Scanning.close_in io)

let scan path =
  try scanning path @@ fun io -> Scanf.bscanf io 
    {|MemTotal: %_d %_s MemFree: %_d %_s MemAvailable: %Ld %s|}
    (fun size kb -> ok (size, kb))
  with _ -> error "failed to scan %s" path

let mem_available () =
  let* (size, kb) = scan "/proc/meminfo" in
  match kb with
  | "kB" -> ok Int64.(mul size 1024L) 
  | kb -> error "unexpected unit: %s" kb
```

File: ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml [hunk 4]
```diff
@@ -545,6 +557,29 @@ let bytes_per_mem_vm = 1024
...
+            read_lines ic
+      else
+        read_lines ic
+    with End_of_file -> close_in ic ; Int64.zero
+  in
+  try
+    let ic = open_in meminfo_file in
+    let value = read_lines ic in
+    close_in ic ; value
+  with Sys_error _ -> Int64.zero
```
Comment (DeliZhangX) [inline ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml:581]: If failed to read `MemAvailable`, I suggest just stop collecting current data so that `memory_internal_free` will be none (unavailable).

File: ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml [hunk 5]
```diff
@@ -580,7 +615,14 @@ let dss_mem_vms doms =
       in
       let other_ds =
         if domid = 0 then
-          None
+          let free_mem = read_mem_available () in
+          Some
+            ( Rrd.VM uuid
+            , Ds.ds_make ~name:"memory_internal_free" ~units:"KiB"
+                ~description:"Dom 0 current free memory to use"
```
Comment (DeliZhangX) [inline ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml:622]: Can be "Dom0 free memory"

File: ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml (hunk 4 above)
Comment (LunfanZhang) [reply ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml:581]: Maybe return **None** as before?

File: ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml [hunk 6]
```diff
@@ -545,6 +557,29 @@ let bytes_per_mem_vm = 1024
...
+            Scanf.sscanf value " %Ld kB" (fun v -> v)
+        | _ ->
+            read_lines ic
+      else
+        read_lines ic
+    with End_of_file -> close_in ic ; Int64.zero
+  in
+  try
+    let ic = open_in meminfo_file in
+    let value = read_lines ic in
```
Comment (liulinC) [inline ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml:579]: If error raised during read_lines, will the FD be closed?
Suggest to investigate whether `finally` can resolve if any issue.

File: ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml [hunk 7]
```diff
@@ -545,6 +557,29 @@ let bytes_per_mem_vm = 1024
...
+        let parts = String.split_on_char ':' line in
+        match parts with
+        | key :: value :: _ when String.trim key = "MemAvailable" ->
+            (* Extract the value, remove the 'kB' suffix, trim spaces, and convert to Int64 *)
+            Scanf.sscanf value " %Ld kB" (fun v -> v)
+        | _ ->
+            read_lines ic
+      else
+        read_lines ic
+    with End_of_file -> close_in ic ; Int64.zero
```
[the remark this answers is not in the document]
Comment (LunfanZhang) [reply ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml:575]: Compared to other metrics, ` None`  will be better.

File: ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml [hunk 8]
```diff
@@ -580,7 +620,17 @@ let dss_mem_vms doms =
...
+          match mem_available () with
+          | Ok mem ->
+              Some
+                ( Rrd.VM uuid
+                , Ds.ds_make ~name:"memory_internal_free" ~units:"B"
+                    ~description:"Dom0 current free memory"
+                    ~value:(Rrd.VT_Int64 mem) ~ty:Rrd.Gauge ~min:0.0
+                    ~default:true ()
+                )
+          | Error _ ->
```
Comment (lindig) [inline ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml:632]: Any error should be logged. The error to log is in this value.

File: ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml [hunk 9]
```diff
@@ -545,6 +557,34 @@ let bytes_per_mem_vm = 1024
 
 let mem_vm_writer_pages = ((max_supported_vms * bytes_per_mem_vm) + 4095) / 4096
 
+let res_error fmt = Printf.kprintf (fun msg -> Result.error (`Msg msg)) fmt
```
Comment (lindig) [inline ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml:560]: The `Msg` part is not needed - it was in my code because I just copied it from somewhere else. The error is a simple string.

**Review by lindig — APPROVED** [review]
I'm approving this now but please add the error logging.

File: ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml (hunk 9 above)
Comment (LunfanZhang) [reply ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml]: Fixed

File: ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml (hunk 8 above)
Comment (LunfanZhang) [reply ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml]: error log was added.

File: ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml [hunk 10]
```diff
@@ -580,7 +620,18 @@ let dss_mem_vms doms =
...
+          match mem_available () with
+          | Ok mem ->
+              Some
+                ( Rrd.VM uuid
+                , Ds.ds_make ~name:"memory_internal_free" ~units:"B"
+                    ~description:"Dom0 current free memory"
+                    ~value:(Rrd.VT_Int64 mem) ~ty:Rrd.Gauge ~min:0.0
+                    ~default:true ()
+                )
+          | Error exn ->
```
Comment (minglumlu) [inline ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml:632]: I think `exn` is the abbreviation of `exception`.

Comment (lindig) [reply ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml:632]: `exn` does not have to be an exception - it is just a string; you could name it `msg` (message) and it would be as valid. We simply want to log it.

File: ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml [hunk 11]
```diff
@@ -580,7 +620,18 @@ let dss_mem_vms doms =
...
+              let _ = error "record Dom0 free memory failed: %s" exn in
```
[the remark this answers is not in the document]
Comment (lindig) [reply ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml:633]: It's always a good idea to log the location: 
```
error "%s: retrieving Dom0 free memory failed: %s" __FUNCTION__ exn
```

**Review by lindig — APPROVED** [review]

File: ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml (hunk 11 above)
[the remark this answers is not in the document]
Comment (LunfanZhang) [reply ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml]: I have added the location.

File: ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml (hunk 10 above)
Comment (LunfanZhang) [reply ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml]: Avoid confusion, change `exn` to `msg `

**Review by minglumlu — APPROVED** [review]
